### ISSUE TITLE
About the RegularLoss (Eq. (10) in the paper).

### ISSUE THREAD

**Issue by xl814** [decisionintelligence/CATCH#16]
Thanks for your excellent work.

Regarding the implementation of **RegularLoss**, I noticed two differences between the paper and the released code:

1. The coefficient of the RegularLoss is set to $$\frac{1}{N(N-1)}$$, whereas it is written as  $$\frac{1}{N}$$ in the paper.

2. The implementation uses the L1 norm, while the paper adopts the Frobenius norm.

Could you please clarify the motivation behind these design choices?

**Rerferenace**: 

https://github.com/decisionintelligence/CATCH/blob/3647c69be5eb56649b072596cf89098e689e20c3/ts_benchmark/baselines/catch/utils/ch_discover_loss.py#L35

<img width="2071" height="391" alt="Image" src="https://github.com/user-attachments/assets/ef16d0b3-6848-4383-aa08-d15344e07a73" />

**Comment by ccloud0525** [comment]
> Thanks for your excellent work.
> 
> Regarding the implementation of **RegularLoss**, I noticed two differences between the paper and the released code:
> 
> 1. The coefficient of the RegularLoss is set to 
>      
>        1
>        
>          N
>          (
>          N
>          −
>          1
>          )
>        
>      
>    , whereas it is written as  
>      
>        1
>        N
>      
>     in the paper.
> 2. The implementation uses the L1 norm, while the paper adopts the Frobenius norm.
> 
> Could you please clarify the motivation behind these design choices?
> 
> **Rerferenace**:
> 
> [CATCH/ts_benchmark/baselines/catch/utils/ch_discover_loss.py](https://github.com/decisionintelligence/CATCH/blob/3647c69be5eb56649b072596cf89098e689e20c3/ts_benchmark/baselines/catch/utils/ch_discover_loss.py#L35)
> 
> Line 35 in [3647c69](/decisionintelligence/CATCH/commit/3647c69be5eb56649b072596cf89098e689e20c3)
> 
>  regular_loss = 1 / (n_vars * (n_vars - 1)) * torch.norm(eye.reshape(b, -1) - attn_mask.reshape((b, -1)), 
> <img alt="Image" width="2000" height="391" src="https://private-user-images.githubusercontent.com/45684305/532720317-ef16d0b3-6848-4383-aa08-d15344e07a73.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc5NTE4NzgsIm5iZiI6MTc2Nzk1MTU3OCwicGF0aCI6Ii80NTY4NDMwNS81MzI3MjAzMTctZWYxNmQwYjMtNjg0OC00MzgzLWFhMDgtZDE1MzQ0ZTA3YTczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA5VDA5MzkzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyYjNlZjgwMGJkN2JmMGIyZTg1ZTg5NGUwODY3MTYzMGZkNjFkNWQ5YmM0Mzg0NjRiZTAxMjE0NzQ2YjA5YTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WlUtjrnmwh4GNl_4lJ4kMMpqKe_v2t4ap7c5U2eYPts">

Thank you for your question. These two implementation methods are actually similar in effect. Assuming that the number of 1s and -1s in the matrix after I-M is X, the implementation in the code is X/N(N-1), while that in the paper is sqrt(X)/N, which is close to the square root form of the former. The paper uses a simpler form for clarity and ease of understanding. Since additional coefficients are added when integrating the values of all loss functions,  the differences will not be particularly significant.

**Comment by xl814** [comment]
Thank you for your reply.

However, it is worth noting that when 𝑁 equals 1, this setting will lead to a division-by-zero error.

**Comment by ccloud0525** [comment]
> Thank you for your reply.
> 
> However, it is worth noting that when 𝑁 equals 1, this setting will lead to a division-by-zero error.

Thanks for your kind reminder. Since CATCH is built for *multivariate* anomaly detection and the mask is always between CI and CD, where the N diagonal elements are always 1s, we divide by N(N - 1). Since the design motivation of CATCH does not consider univariate anomaly detection, we did not notice this "division-by-zero" issue.
